internals/overlord/logstate/syslog/syslog_test.go

```diff
@@ -0,0 +1,295 @@
+
+	flushReturned := make(chan struct{})
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+		defer cancel()
+		err = client.Flush(ctx)
+		c.Check(err, NotNil)
+		close(flushReturned)
+	}()
+
+	// Check Flush returns quickly after context timeout
+	select {
+	case <-flushReturned:
+	case <-time.After(1 * time.Second):
+		c.Fatal("lokiClient.Flush took too long to return after context timeout")
+	}
+
+}
+func (*suite) TestBufferFull(c *C) {
+	client, err := syslog.NewClient(&syslog.ClientOptions{
+		TargetName:        "tgt1",
+		Location:          "tcp://fake:514",
+		MaxRequestEntries: 3,
+	})
+	c.Assert(err, IsNil)
+
+	addEntry := func(s string) {
+		err := client.Add(servicelog.Entry{Message: s})
+		c.Assert(err, IsNil)
+	}
+
+	// Check that the client's buffer is as expected
+	buffer := syslog.GetBuffer(client)
+	checkBuffer := func(expected []any) {
+		if len(buffer) != len(expected) {
+			c.Fatalf("buffer length is %v, expected %v", len(buffer), len(expected))
+		}
+
+		for i := range expected {
+			// 'nil' means c.buffer[i] should be zero
+			if expected[i] == nil {
+				c.Assert(buffer[i], DeepEquals, syslog.EntryWithService{},
+					Commentf("buffer[%d] should be zero, obtained %v", i, buffer[i]))
+				continue
+			}
+
+			// Otherwise, check buffer message matches string
+			msg := expected[i].(string)
+			c.Assert(syslog.GetMessage(buffer[i]), Equals, msg)
+		}
+	}
+
+	checkBuffer([]any{nil, nil, nil, nil, nil, nil})
+	addEntry("1")
+	checkBuffer([]any{"1", nil, nil, nil, nil, nil})
+	addEntry("2")
+	checkBuffer([]any{"1", "2", nil, nil, nil, nil})
+	addEntry("3")
+	checkBuffer([]any{"1", "2", "3", nil, nil, nil})
+	addEntry("4")
+	checkBuffer([]any{nil, "2", "3", "4", nil, nil})
+	addEntry("5")
+	checkBuffer([]any{nil, nil, "3", "4", "5", nil})
+	addEntry("6")
+	checkBuffer([]any{nil, nil, nil, "4", "5", "6"})
+	addEntry("7")
+	checkBuffer([]any{"5", "6", "7", nil, nil, nil})
+}
+
+func (*suite) TestFlushEmpty(c *C) {
+	client, err := syslog.NewClient(&syslog.ClientOptions{
+		Location:   "tcp://fake:514",
+		TargetName: "test",
+	})
+	c.Assert(err, IsNil)
+
+	// Flushing with no entries should be a no-op
+	err = client.Flush(context.Background())
+	c.Assert(err, IsNil)
+}
+
+func (*suite) TestInvalidLocation(c *C) {
+	// Invalid scheme
+	_, err := syslog.NewClient(&syslog.ClientOptions{
+		Location: "http://example.com:514",
+	})
+	c.Assert(err, ErrorMatches, `invalid syslog server address scheme "http"`)
+
+	// Valid schemes should work
+	_, err = syslog.NewClient(&syslog.ClientOptions{
+		Location: "tcp://localhost:514",
+	})
+	c.Assert(err, IsNil)
+
+	_, err = syslog.NewClient(&syslog.ClientOptions{
+		Location: "udp://localhost:514",
+	})
+	c.Assert(err, IsNil)
+}
+
+type testSyslogServer struct {
+	listener net.Listener
+	msgChan  chan string
+}
+
+func (s *testSyslogServer) close() error {
+	return s.listener.Close()
+}
+
+func (s *testSyslogServer) run() error {
```

Comment:
Ideally we can simplify this similar to https://github.com/canonical/pebble/pull/165/files#diff-4381a454fdf2083e6046457179814146e2d8465dbd33ca69d4fcd0e4458c663eR153